plutus-core/plutus-ir/Language/PlutusIR/Transform/Inline.hs

```diff
@@ -0,0 +1,245 @@
+preserve it because we don't currently inline anything with binders.
+
+If we do start inlining things with binders in them, we should probably try and preserve it by
+doing something like 'The rapier' section from the paper.
+-}
+
+-- | Inline simple bindings. Relies on global uniqueness, and preserves it.
+-- See Note [Inlining and global uniqueness]
+inline
+    :: ExternalConstraints tyname name uni a term
+    => DynamicBuiltinNameMeanings term
+    -> Term tyname name uni a
+    -> Term tyname name uni a
+inline means t =
+    let
+        -- We actually just want the variable strictness information here!
+        deps :: (G.Graph Deps.Node, Map.Map PLC.Unique Strictness)
+        deps = Deps.runTermDeps means t
+    in flip runReader (snd deps, means) $ flip evalStateT mempty $ processTerm t
+
+{- Note [Removing inlined bindings]
+We *do* remove bindings that we inline (since we only do unconditional inlining). We *could*
+leave this to the dead code pass, but we might as well do it here, since it's a very similar decision.
+Crucially, we have to do the same reasoning wrt strict bindings and purity (see Note [Inlining and purity]):
+we can only inline *pure* strict bindings, which is effectively the same as what we do in the dead
+code pass.
+
+Annoyingly, this requires us to redo the analysis that we do for the dead binding pass.
+TODO: merge them or figure out a way to share more work, especially since there's similar logic.
+This might mean reinventing GHC's OccAnal...
+-}
+
+processTerm
+    :: Inlining tyname name uni a term m
+    => Term tyname name uni a
+    -> m (Term tyname name uni a)
+processTerm = \case
+    v@(Var _ n) -> do
+        subst <- get
+        case lookupSubst n subst of
+            -- Not substituted for, leave it as it is
+            Nothing -> pure v
+            -- Already processed term, just put it in, don't do any further optimization here.
+            -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+            Just (Done t) -> pure t
+    Let a NonRec bs t -> do
+        -- Process bindings, eliminating those which will be inlined unconditionally,
+        -- and accumulating the new substitutions
+        -- See Note [Removing inlined bindings]
+        bs' <- wither processSingleBinding (toList bs)
+        t' <- processTerm t
+        -- Use 'mkLet': we're using lists of bindings rather than NonEmpty since we might actually
+        -- have got rid of all of them!
+        pure $ mkLet a NonRec bs' t'
+    -- This includes recursive let terms, we don't even consider inlining them at the moment
+    t -> forMOf termSubterms t processTerm
+
+{- Note [Inlining various kinds of binding]
+We can inline term and type bindings, we can't do anything with datatype bindings.
+
+We don't actually inline type bindings at the moment, mostly because I think it
+won't get us much as they aren't created very often.
+-}
```

Comment:
Presumably we're mostly interested in run-time performance here, so if we're going to throw away the types anyway there's not much to be gained by optimising them.